src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs

```diff
@@ -12,6 +12,48 @@ namespace System.Collections.Immutable
     /// </summary>
     public static class ImmutableInterlocked
     {
+        /// <summary>
+        /// Mutates a value in-place with optimistic locking transaction semantics
+        /// via a specified transformation function.
+        /// The transformation is retried as many times as necessary to win the optimistic locking race.
+        /// </summary>
+        /// <typeparam name="T">The type of data.</typeparam>
+        /// <param name="hotLocation">
+        /// The variable or field to be changed, which may be accessed by multiple threads.
+        /// </param>
+        /// <param name="transformation">
+        /// A function that mutates the value. This function should be side-effect free, 
+        /// as it may run multiple times when races occur with other threads.</param>
+        /// <returns>
+        /// <c>true</c> if the location's value is changed by applying the result of the 
+        /// <paramref name="transformation"/> function;
+        /// <c>false</c> if the location's value remained the same because the last 
+        /// invocation of <paramref name="transformation"/> returned the existing value.
+        /// </returns>
+        public static bool ApplyChange<T>(ref T hotLocation, Func<T, T> transformation) where T : class
```

Comment:
Sorry, I sent the pull request before applying some of the API review feedback. I've pushed commits to fix both of those issues.